ahcore/callbacks/milvus_vector_db_callback.py

```diff
@@ -0,0 +1,111 @@
+            and data preparation. E.g. "path_fomo_cls" for a pathology fomo with Image metadata and CLS embeddings.
+        embedding_dim: int
+            The output dimension of your embedding model
+        database : str
+            The database type to connect to. Default is "default".
+        """
+        self.host = host
+        self.port = int(port)
+        self.user = user
+        self.database = database
+        self._password = password
+        self._connection_alias = "ahcore_milvus_vector_db"
+        connections.connect(
+            host=self.host, port=self.port, user=self.user, password=self._password, alias=self._connection_alias
+        )
+
+        self._collection_name = collection_name
+        self._collection_type = CollectionType[collection_type.upper()]
+        self._collection: Collection | None = None
+        self._prepare_data: Callable[
+            [Dict[str, Any], Dict[str, Any]], PrepareDataTypePathCLS | PrepareDataTypeDebug
+        ] | None = None
+
+        self.embedding_dim = embedding_dim
+
+    def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
+        try:
+            self._collection = Collection(self._collection_name, using=self._connection_alias)
+        except MilvusException:
+            logger.info(f"Collection {self._collection_name} does not exist. Creating it now.")
+            factory = MilvusCollectionFactory(embedding_dim=self.embedding_dim)
+            schema = factory.get_schema(self._collection_type)
+            self._collection = Collection(self._collection_name, schema, using=self._connection_alias)
+
+        prepare_data_factory: PrepareDataFactory = PrepareDataFactory()
+        self._prepare_data = prepare_data_factory.get_preparer(self._collection_type)
+
+    @property
+    def collection(self) -> Collection:
+        return self._collection
+
+    def on_predict_batch_end(
+        self,
+        trainer: Trainer,
+        pl_module: LightningModule,
+        outputs: dict[str, Any],
+        batch: dict[str, Any],
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        # for mypy
+        assert self._prepare_data, "prepare_data should be set in the setup method."
+        assert self._collection, "collection should be set in the setup method."
+
+        prepared_data = list(self._prepare_data(outputs, batch))  # convert to list for milvus
+
+        # Insert data into Milvus collection
+        try:
+            self._collection.insert(prepared_data)
+        except MilvusException as e:
+            logger.error(f"Error inserting data into Milvus collection: {e}")
+
+    def on_predict_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        self.collection.flush()  # this is last flush, milvus auto-flushes whenever reaching a certain cache size
+        connections.disconnect(self._connection_alias)
```

Comment:
Will the connection still work when new epoch?

Reply:
There is no new epoch, this is predict only, which will only run once on the provided data (inference)